### ISSUE TITLE
gan loss only discriminator loss is used

### ISSUE THREAD

**Issue by prabakar2532002** [microsoft/InvisMark#11]
In the training code you have used gan loss, where 
```self.losses['discriminator_loss'] += real_loss + fake_loss```
this discriminator loss is used in the training code where all of the models are updated with this loss (since this loss tends to increase the difference between the watermarked image and original cover image) updating the encoder model parameters with this loss leads to unexpected results.
in w-gan loss generator and discriminator losses will be calculated as below
```
discriminator loss = fake_loss - real_loss     # Objective - maximise the difference between watermarked image and cover image
generator loss = -fake_loss       # increase the value of fake loss so that it increases the quality of watermarked image
```
since both has opposing objectives we need to update the weights separately in torch (discriminator loss has to update the weight of discriminator model only and generator loss has to update the generator model only (encoder))
****
unlike the above w-gan example the training code uses only discriminator loss and all of the models are updated with the discriminator loss isn't this will oppose the model learning purpose, someone please let me know whether my argument is correct or is there anything i am missing.

Thank you.
